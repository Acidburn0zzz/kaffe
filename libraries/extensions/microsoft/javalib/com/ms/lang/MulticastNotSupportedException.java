package com.ms.lang;

import java.lang.Exception;

public class MulticastNotSupportedException extends Exception {

public MulticastNotSupportedException() {
	super();
}

public MulticastNotSupportedException(String mess) {
	super(mess);
}

}
