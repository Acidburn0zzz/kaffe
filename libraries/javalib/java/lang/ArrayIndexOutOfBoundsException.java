package java.lang;


/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */
public class ArrayIndexOutOfBoundsException
  extends IndexOutOfBoundsException
{
public ArrayIndexOutOfBoundsException () {
	super();
}

public ArrayIndexOutOfBoundsException (String s) {
	super(s);
}

public ArrayIndexOutOfBoundsException (int index) {
	super(String.valueOf(index));
}
}
