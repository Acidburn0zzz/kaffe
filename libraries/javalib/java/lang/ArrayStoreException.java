package java.lang;


/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */
public class ArrayStoreException
  extends RuntimeException
{
public ArrayStoreException () {
	super();
}

public ArrayStoreException (String s) {
	super(s);
}
}
