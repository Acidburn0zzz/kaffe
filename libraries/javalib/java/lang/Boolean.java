/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */

package java.lang;

import java.io.Serializable;

final public class Boolean implements Serializable {

public static final Boolean TRUE = new Boolean(true);
public static final Boolean FALSE = new Boolean(false);
public static final Class TYPE = Class.getPrimitiveClass("boolean");

/* This is what Sun's JDK1.1 "serialver java.lang.Boolean" spits out */
private static final long serialVersionUID = -3665804199014368530L;

private final boolean value;

public Boolean(String s) {
	value = (s != null && s.toLowerCase().equals("true"));
}

public Boolean(boolean value) {
	this.value=value;
}

public boolean booleanValue() {
	return (value);
}

public boolean equals(Object obj) {
	return (obj instanceof Boolean)
	  && (((Boolean) obj).value == this.value);
}

public static boolean getBoolean(String name) {
	String value = System.getProperty(name);
	if (value == null) {
		return (false);
	}
	else {
		return (value.equalsIgnoreCase("true"));
	}
}

public int hashCode() {
	return (value ? 1231 : 1237);
}

public String toString() {
	return (value ? "true" : "false");
}

public static Boolean valueOf(String s) {
	return new Boolean(s);
}
}
