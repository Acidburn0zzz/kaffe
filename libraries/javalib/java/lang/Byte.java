package java.lang;


/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */
final public class Byte extends Number
{
	public static final byte MIN_VALUE = -0x80;
	public static final byte MAX_VALUE = 0x7F;
	public static final Class TYPE = Class.getPrimitiveClass("byte");
	private final byte value;

public Byte(String s) throws NumberFormatException {
	this(parseByte(s));
}

public Byte(byte value) {
	this.value = value;
}

public static Byte decode(String nm) throws NumberFormatException {
	byte val;

	if (nm.value[nm.offset] == '#') {
		val = parseByte(nm.substring(1), 16);
	}
	else if (nm.value[nm.offset] == '0') {
		if ( (nm.count > 1) && (nm.value[nm.offset+1] == 'x') )
			val = parseByte( nm.substring(2), 16);
		else
			val = parseByte(nm.substring(1), 8);
	}
	else {
		val = parseByte(nm, 10);
	}

	return new Byte(val);
}

public double doubleValue() {
	return ((double)value);
}

public boolean equals(Object obj) {
	return (obj != null)
	    && (obj instanceof Byte)
	    && (((Byte) obj).value == this.value);
}

public float floatValue() {
	return ((float)value);
}

public int hashCode() {
	return (value);	// What should this be do you suppose ???
}

public byte byteValue() {
	return (value);
}

public short shortValue() {
	return ((short)value);
}

public int intValue() {
	return ((int)value);
}

public long longValue() {
	return ((long)value);
}

public static byte parseByte(String s) throws NumberFormatException {
	return (parseByte(s, 10));
}

public static byte parseByte(String s, int radix) throws NumberFormatException {
	return ((byte)Integer.parseInt(s, radix));
}

public String toString() {
	return (toString(value));
}

public static String toString(byte s) {
	return Integer.toString((int)s);
}

public static Byte valueOf(String s) throws NumberFormatException {
	return (new Byte(parseByte(s)));
}

public static Byte valueOf(String s, int radix) throws NumberFormatException {
	return (new Byte(parseByte(s, radix)));
}
}
