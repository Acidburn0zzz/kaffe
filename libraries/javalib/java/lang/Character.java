/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */

// This is only a partial implementation of the JDK 1.1 spec.  We need
// to do all the necessary work to get Unicode happening - this is just
// an ASCII subset.

package java.lang;

public final class Character extends Object {

  public static final int MIN_RADIX = 2;
  public static final int MAX_RADIX = 36;
  public static final char MIN_VALUE = 0x0000;
  public static final char MAX_VALUE = 0xffff;
  public static final int UNASSIGNED = 0;
  public static final int UPPERCASE_LETTER = 1;
  public static final int LOWERCASE_LETTER = 2;
  public static final int TITLECASE_LETTER = 3;
  public static final int MODIFIER_LETTER = 4;
  public static final int OTHER_LETTER = 5;
  public static final int NON_SPACING_MARK = 6;
  public static final int ENCLOSING_MARK = 7;
  public static final int COMBINING_SPACING_MARK = 8;
  public static final int DECIMAL_DIGIT_NUMBER = 9;
  public static final int LETTER_NUMBER = 10;
  public static final int OTHER_NUMBER = 11;
  public static final int SPACE_SEPARATOR = 12;
  public static final int LINE_SEPARATOR = 13;
  public static final int PARAGRAPH_SEPARATOR = 14;
  public static final int CONTROL = 15;
  public static final int FORMAT = 16;
  public static final int PRIVATE_USE = 18;
  public static final int SURROGATE = 19;
  public static final int DASH_PUNCTUATION = 20;
  public static final int START_PUNCTUATION = 21;
  public static final int END_PUNCTUATION = 22;
  public static final int CONNECTOR_PUNCTUATION = 23;
  public static final int OTHER_PUNCTUATION = 24;
  public static final int MATH_SYMBOL = 25;
  public static final int CURRENCY_SYMBOL = 26;
  public static final int MODIFIER_SYMBOL = 27;
  public static final int OTHER_SYMBOL = 28;

  public static final Class TYPE = Class.getPrimitiveClass("char");

  private final char value;

  public Character(char value)
  {
    this.value=value;
  }

  public char charValue()
  {
    return (value);
  }

  public int hashCode()
  {
    return ((int)value);
  }

  public boolean equals(Object obj)
  {
    if (obj instanceof Character) {
      return (charValue() == ((Character)obj).charValue());
    }
    else {
      return (false);
    }    
  }

  public String toString()
  {
    return (String.valueOf(value));
  }

  public static boolean isSpace(char ch)
  {
    return (isWhitespace(ch));
  }

  public static boolean isLetterOrDigit(char ch)
  {
    return (isLetter(ch) || isDigit(ch));
  }
  
  public static boolean isJavaLetter(char ch)
  {
    return (isJavaIdentifierStart(ch));
  }
  
  public static boolean isJavaLetterOrDigit(char ch)
  {
    return (isJavaIdentifierPart(ch));
  }

  public static boolean isTitleCase(char ch)
  {
    return (ch == 0x01C5 || ch == 0x01C8 || ch == 0x01CB || ch == 0x01F2);
  }
  
  public static char toTitleCase(char ch)
  {
    if (ch == 0x01C6) {
      return (0x01C6);
    }
    else if (ch == 0x01C9) {
      return (0x01C8);
    }
    else if (ch == 0x01CC) {
      return (0x01CB);
    }
    else if (ch == 0x01F3) {
      return (0x01F2);
    }
    else {
      return (ch);
    }
  }

  /**
   * Determines if a character has a defined meaning in Unicode. 
   * A character is defined if at least one of the following is true:  <br>
   *   It has an entry in the Unicode attribute table.  <br>
   *   Its value is in the range 0x3040 <= ch <= 0x9FA5.  <br>
   *   Its value is in the range 0xF900 <= ch <= 0xFA2D.  <br>
   */
  public static boolean isDefined(char ch) {
	if (0x3040 <= ch && ch <= 0x9FA5 ||
	    0xF900 <= ch && ch <= 0xFA2D) {
		return (true);
	}
	/* FIXME: we do not check for entry in Unicode attribute table.
	 * See comment at beginning of file that this impl is partial.
	 */
	return (ch < 256);
  }

  public static boolean isIdentifierIgnorable(char ch)
  {
    if ((ch >= 0x0000 && ch <= 0x0008) ||
        (ch >= 0x000E && ch <= 0x001B) ||
        (ch >= 0x007F && ch <= 0x009F) ||
        (ch >= 0x200C && ch <= 0x200F) ||
        (ch >= 0x202A && ch <= 0x202E) ||
        (ch >= 0x206A && ch <= 0x206F) ||
        (ch == 0xFEFF)) {
      return (true);
    }
    else {
      return (false);
    }
  }

  public static boolean isLowerCase(char ch)
  {
    return (ch >= 'a' && ch <= 'z');
  }

  public static boolean isUpperCase(char ch)
  {
    return (ch >= 'A' && ch <= 'Z');
  }

  public static boolean isDigit(char ch)
  {
    return (ch >= '0' && ch <= '9');
  }

  public static boolean isLetter(char ch)
  {
    return ((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z'));
  }

  public static char toLowerCase(char ch)
  {
    char result=ch;

    if (isUpperCase(ch)) result=(char )(((int )ch)+32);

    return result;
  }

  public static char toUpperCase(char ch)
  {
    char result=ch;

    if (isLowerCase(ch)) result=(char )(((int )ch)-32);

    return result;
  }

  public static int digit(char ch, int radix)
  {
    if (radix < MIN_RADIX || radix > MAX_RADIX) {
      return -1;
    }

    int val = radix;
    if (isDigit(ch)) {
      val = ch - '0';
    }
    else if (isLetter(ch)) {
      val = ((int)toLowerCase(ch)) - ((int)'a') + 10;
    }

    if (val < radix) {
      return (val);
    }
    else {
      return (-1);
    }
  }

  public static char forDigit(int digit, int radix)
  {
    if ((radix<MIN_RADIX) || (radix>MAX_RADIX)) return 0x0000;
    if (digit>radix) return 0x0000;

    if (digit<10) {
      return (char )(((int )'0')+digit);
    }
    else {
      return (char )(((int )'a')+(digit-10));
    }
  }

  public static int getNumericValue(char ch)
  {
    return (digit(ch, 10));
  }

  public static int getType(char ch)
  {
    return (UNASSIGNED);
  }

  public static boolean isISOControl(char ch)
  {
    if ((ch >= 0x0000 && ch <= 0x001F) ||
        (ch >= 0x007F && ch <= 0x009F)) {
      return (true);
    }
    else {
      return (false);
    }
  }

  public static boolean isJavaIdentifierPart(char ch)
  {
      return ((ch >= 0x0000 && ch <= 0x0008) ||
	(ch >= 0x000e && ch <= 0x001b) ||
	(ch >= 0x007f && ch <= 0x009f) ||
	ch == '$' || ch == '_' ||
        isLetter(ch) || isDigit(ch));
  }

  public static boolean isJavaIdentifierStart(char ch)
  {
    return (isLetter(ch) || ch == '$' || ch == '_');
  }

  public static boolean isSpaceChar(char ch)
  {
      return (ch == ' ');
  }

  public static boolean isUnicodeIdentifierPart(char ch)
  {
      return (ch != '$' && (isLetter(ch) || isJavaIdentifierPart(ch)));
  }

  public static boolean isUnicodeIdentifierStart(char ch)
  {
      return (isLetter(ch));
  }

  public static boolean isWhitespace(char ch)
  {
    return((ch == ' ')
	|| (ch >= 0x0009 && ch <= 0x000d)
	|| (ch >= 0x001c && ch <= 0x001f) );
  }
}

