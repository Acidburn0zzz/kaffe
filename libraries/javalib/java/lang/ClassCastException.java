package java.lang;


/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */
public class ClassCastException
  extends RuntimeException
{
public ClassCastException () {
	super();
}

public ClassCastException (String s) {
	super(s);
}
}
