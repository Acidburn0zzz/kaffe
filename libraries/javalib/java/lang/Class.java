/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */

package java.lang;

import java.io.InputStream;
import java.io.Serializable;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Member;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.net.URL;
import java.util.Vector;
import kaffe.lang.SystemClassLoader;

public final class Class implements Serializable {

/* For GCJ compatibility, we cannot define any fields in
 * java.lang.Class at this point.  Valid as of 10/28/99
 * We special case it instead in clib/native/ObjectStreamClassImpl.c
 * --gback
 */
// private static final long serialVersionUID = 3206093459760846163L;

// Only instantiable by the VM
private Class() {
}

/*
 * We must use the ClassLoader associated with the calling method/class.
 * We must handle the special case of code like this:
 *
 *    Class c = Class.class;
 *    Method m = c.getMethod("forName", new Class[] { String.class });
 *    c = (Class)m.invoke(c, new Object[] { "Class2" });
 *
 * If we didn't, then we would detect java.lang.reflect.Method as the
 * calling class, and end up always using the bootstrap ClassLoader.
 * To deal with this, we skip over java.lang.reflect.Method.
 */
public static Class forName(String className) throws ClassNotFoundException {
	Class callingClass = getStackClass(1);
	if (callingClass != null
	    && callingClass.getName().equals("java.lang.reflect.Method"))
		callingClass = getStackClass(2);
	return forName(className, true,
	    callingClass == null ? null : callingClass.getClassLoader());
}

public static native Class forName(String className,
	boolean initialize, ClassLoader loader) throws ClassNotFoundException;

private String fullResourceName(String name) {
	if (name.charAt(0) == '/') {
		return (name.substring(1));
	}
	String cname = getName();
	StringBuffer buf = new StringBuffer();
	int tail = cname.lastIndexOf('.');
	if (tail != -1) {
		buf.append(cname.substring(0, tail+1).replace('.', '/'));
	}
	buf.append(name);
	return (buf.toString());
}

/**
 * Determines the class loader for the class.
 *
 * @returns   the class loader that created the class or interface
 *            represented by this object, or null if the class was not
 *            created by a class loader.
 * @see       java.lang.ClassLoader
 */
public ClassLoader getClassLoader() {
	return (getClassLoader0());
}

native private ClassLoader getClassLoader0();

/**
 * If this is an inner class, return the class that
 * declared it.  If not, return null.
 *
 * @return the declaring class of this class.
 * @since JDK1.1
 */
public Class getDeclaringClass() {
	Class[] classes = getClasses0(false);

	switch (classes.length) {
	case 0:
		return null;
	case 1:
		return classes[0];
	default:
		throw new ClassFormatError ("Too many outer classes :" + classes.length);
	}
}

/**
 * Get all the public inner classes, declared in this
 * class or inherited from superclasses, that are
 * members of this class.
 *
 * @return all public inner classes in this class.
 * @since JDK1.1
 */
public Class[] getClasses() {
	SecurityManager sm = System.getSecurityManager();
	Vector v = new Vector();
	Class clazz = this;
	while (clazz != null) {
		if (sm != null)
			sm.checkMemberAccess(clazz, Member.PUBLIC);
		Class[] classes = clazz.getClasses0(true);

		for (int i = 0; i < classes.length; i++) {
			if (Modifier.isPublic(classes[i].getModifiers())) {
				v.add (classes[i]);
			}
		}
		clazz = clazz.getSuperclass();
	}

	return (Class[])v.toArray (new Class[v.size()]);
}

native private Class[] getClasses0(boolean inner);

native public Class getComponentType();

public Constructor getConstructor(Class parameterTypes[]) throws NoSuchMethodException, SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.PUBLIC);
	return (getConstructor0(parameterTypes, false));
}

native private Constructor getConstructor0(Class[] args, boolean declared);

public Constructor[] getConstructors() throws SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.PUBLIC);
	return (getConstructors0(false));
}

native private Constructor[] getConstructors0(boolean declared);

/**
 * Get all the inner classes declared in this class.
 *
 * @return all inner classes declared in this class.
 * @exception SecurityException if you do not have access
 *            to non-public inner classes of this class.
 * @since JDK1.1
 */
public Class[] getDeclaredClasses() throws SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.DECLARED);
	return (getClasses0(true));
}

public Constructor getDeclaredConstructor(Class parameterTypes[]) throws NoSuchMethodException, SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.DECLARED);
	return (getConstructor0(parameterTypes, true));
}

public Constructor[] getDeclaredConstructors() throws SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.DECLARED);
	return (getConstructors0(true));
}

public Field getDeclaredField(String name) throws NoSuchFieldException, SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.DECLARED);
	return (getField0(name, true));
}

public Field[] getDeclaredFields() throws SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.DECLARED);
	return (getFields0(true));
}

public Method getDeclaredMethod(String name, Class parameterTypes[]) throws NoSuchMethodException, SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.DECLARED);
	return (getMethod0(name, parameterTypes, true));
}

public Method[] getDeclaredMethods() throws SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.DECLARED);
	return (getMethods0(true));
}

public Field getField(String name) throws NoSuchFieldException, SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.PUBLIC);
	return (getField0(name, false));
}

native private Field getField0(String name, boolean declared);

public Field[] getFields() throws SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.PUBLIC);
	return (getFields0(false));
}

native private Field[] getFields0(boolean declared);

native public Class[] getInterfaces();

public Method getMethod(String name, Class parameterTypes[]) throws NoSuchMethodException, SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.PUBLIC );
	return (getMethod0(name, parameterTypes, false));
}

native private Method getMethod0(String name, Class[] args, boolean declared);

public Method[] getMethods() throws SecurityException
{
	SecurityManager sm = System.getSecurityManager();
	if (sm != null)
		sm.checkMemberAccess(this, Member.PUBLIC);
	return (getMethods0(false));
}

native private Method[] getMethods0(boolean declared);

native public int getModifiers();

native public String getName();

native static Class getPrimitiveClass(String name);

/**
 * Finds a resource with the specified name.
 * The rules for searching for resources associated with a given class
 * are implemented by the class loader of the class.
 * <p>
 * The Class methods delegate to ClassLoader methods, after applying a
 * naming convention: if the resource name starts with "/", it is used as
 * is. Otherwise, the name of the package is prepended, after converting
 * "." to "/".
 *
 * @param 	name the string representing the resource to be found.
 * @return 	the URL object having the specified name, or null if no
 * resource with the specified name is found.  */
public URL getResource(String name) {
	ClassLoader loader = getClassLoader();
	if (loader == null) {
		loader = SystemClassLoader.getClassLoader();
	}
	return (loader.getResource(fullResourceName(name)));
}

/**
 * Finds a resource with a given name.
 * Will return null if no resource with this name is found. The rules
 * for searching a resources associated with a given class are implemented
 * by the ClassLoader of the class.
 *
 * The Class methods delegate to ClassLoader methods, after applying a
 * naming convention: if the resource name starts with "/", it is used
 * as is. Otherwise, the name of the package is prepended, after
 * converting "." to "/".
 *
 * @param 	name the string representing the resource to be found
 * @return 	the InputStream object having the specified name, or null
 *		if no resource with the specified name is found.
 */
public InputStream getResourceAsStream(String name) {
	ClassLoader loader = getClassLoader();
	if (loader == null)  {
		loader = SystemClassLoader.getClassLoader();
	}
	InputStream strm = loader.getResourceAsStream(fullResourceName(name));
	return (strm);
}

native public Object[] getSigners();
native void setSigners(Object[] signers);

native public Class getSuperclass();

native public boolean isArray();

native public boolean isAssignableFrom(Class cls);

native public boolean isInstance(Object obj);

native public boolean isInterface();

native public boolean isPrimitive();

native public Object newInstance() throws InstantiationException, IllegalAccessException;

/*
 * toString()
 */
public String toString() {
	return (isInterface() ? "interface " : isPrimitive() ? "" : "class ")
	    + getName();
}

/*
 * Determine the Class associated with the method N frames up the stack:
 *
 * Frame #      Method
 * -------      ------
 *   -2		SecurityManager.getClassContext0()
 *   -1		Class.getStackClass()
 *    0		The method calling Class.getStackClass()
 *    1		The method calling the method calling Class.getStackClass()
 *    2		...etc...
 *
 * Returns null if not found.
 */
static Class getStackClass(int frame) {
	Class[] classStack = SecurityManager.getClassContext0();
	frame += 2;
	if (frame >= 0 && frame < classStack.length)
		return classStack[frame];
	return null;
}

}
