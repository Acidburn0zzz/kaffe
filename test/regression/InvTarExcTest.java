import java.lang.reflect.*;

public class InvTarExcTest
{
    public static void f() throws Exception {
	throw new Exception("throwing inner exception");
    }

    public static void main(String av[]) {
	try {
	    Class c = Class.forName("InvTarExcTest");
	    Method m = c.getMethod("f", new Class[0]);
	    m.invoke(null, null);
	} catch (InvocationTargetException e) {
	    Throwable t = e.getTargetException();
	    System.out.println("Caught itexception: " + e.getMessage());
	    e.printStackTrace(System.out);
	    System.out.println("Target is: " + t.getMessage());
	    // t.printStackTrace(System.out);
	} catch (Throwable t) {
	    System.out.println("Caught throwable: " + t);
	    t.printStackTrace(System.out);
	    System.out.println("This should not happen.");
	}
    }
}

/* Expected Output:
Caught itexception: null
java.lang.reflect.InvocationTargetException: java.lang.Exception: throwing inner exception
	at java.lang.Throwable.fillInStackTrace(Throwable.java:native)
	at java.lang.Throwable.<init>(Throwable.java:38)
	at java.lang.Exception.<init>(Exception.java:24)
	at InvTarExcTest.f(InvTarExcTest.java:6)
	at java.lang.reflect.Method.invoke0(Method.java:native)
	at java.lang.reflect.Method.invoke(Method.java:259)
	at InvTarExcTest.main(InvTarExcTest.java:13)
Target is: throwing inner exception
*/
