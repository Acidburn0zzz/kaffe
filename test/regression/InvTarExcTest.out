Caught itexception: null
java.lang.reflect.InvocationTargetException
	at InvTarExcTest.main(InvTarExcTest.java:13)
Target is: throwing inner exception
